qCC/ccCommandLineCommands.cpp

```diff
@@ -2688,6 +2690,164 @@ bool CommandMergeClouds::process(ccCommandLineInterface& cmd)
+			cmd.arguments().pop_front();
+
+			keepOrigFixed = true;
+			cmd.print(QObject::tr("[%1]").arg(COMMAND_SET_GLOBAL_SHIFT_KEEP_ORIG_FIXED));
+		}
+		else {
+			break;
+		}
+	}
+
+	//create an entity vector
+	size_t nrOfClouds = cmd.clouds().size();
+	size_t nrOfMeshes = cmd.meshes().size();
+	size_t nrOfEntities = nrOfClouds + nrOfMeshes;
+	QVector<QPair<ccShiftedObject*, CLEntityDesc*>> entities;
+
+	//add clouds to the vector
+	for (size_t i = 0; i < nrOfClouds; i++)
+	{
+		QPair<ccShiftedObject*, CLEntityDesc*> entity;
+		entity.first = cmd.clouds()[i].pc;
+		entity.second = &cmd.clouds()[i];
+		entities.append(entity);
+	}
+
+	//add meshes to the vector
+	for (size_t i = 0; i < nrOfMeshes; i++)
+	{
+		QPair<ccShiftedObject*, CLEntityDesc*> entity;
+		bool isLocked = false;
+		ccShiftedObject* shifted = ccHObjectCaster::ToShifted(cmd.meshes()[i].mesh, &isLocked);
+		if (shifted && !isLocked)
+		{
+			entity.first = shifted;
+			entity.second = &cmd.meshes()[i];
+			entities.append(entity);
+		}
+	}
+
+	//process both clouds and meshes
+	for (size_t i = 0; i < nrOfEntities; i++)
+	{
+		QPair<ccShiftedObject*, CLEntityDesc*> entity=entities[i];
+		CLEntityDesc& desc = *entity.second;
+		ccShiftedObject* shiftedObject = entity.first;
+		CCVector3d originalShift = shiftedObject->getGlobalShift();
+		cmd.print(QObject::tr("\t[%4 - %5] Original global shift {%1,%2,%3}")
+			.arg(originalShift.x)
+			.arg(originalShift.y)
+			.arg(originalShift.z)
+			.arg(desc.basename)
+			.arg(shiftedObject->getName()));
+
+		//translate entity to keep the same origin
+		if (keepOrigFixed)
+		{
+			CCVector3d T = newShift - originalShift;
+			ccGLMatrix transMat;
+			double maxCoordValue = ccGlobalShiftManager::MaxCoordinateAbsValue();
+			if (T.x > maxCoordValue || T.y > maxCoordValue || T.z > maxCoordValue)
+			{
+				cmd.warning(QObject::tr("\t[%5 - %6] Applied transformation is bigger {%1,%2,%3} than the threshold {%4}, precision loss may occure.")
+					.arg(T.x)
+					.arg(T.y)
+					.arg(T.z)
+					.arg(maxCoordValue)
+					.arg(desc.basename)
+					.arg(shiftedObject->getName()));
+			}
+
+			cmd.print(QObject::tr("\t[%4 - %5] Applied Transformation {%1,%2,%3}")
+				.arg(T.x)
+				.arg(T.y)
+				.arg(T.z)
+				.arg(desc.basename)
+				.arg(shiftedObject->getName()));
+			transMat.toIdentity();
+			transMat.setTranslation(T);
+			shiftedObject->applyGLTransformation_recursive(&transMat);
+		}
+
+		//apply new global shift
+		shiftedObject->setGlobalShift(newShift.x, newShift.y, newShift.z);
+		cmd.print(QObject::tr("\t[%4 - %5] Global shift set to {%1,%2,%3}")
+			.arg(newShift.x)
+			.arg(newShift.y)
+			.arg(newShift.z)
+			.arg(desc.basename)
+			.arg(shiftedObject->getName()));
+		QString nameSuffix = QObject::tr("_SHIFTED_FROM_%1_%2_%3_TO_%4_%5_%6")
+			.arg(originalShift.x)
+			.arg(originalShift.y)
+			.arg(originalShift.z)
+			.arg(newShift.x)
+			.arg(newShift.y)
+			.arg(newShift.z);
+		if ((&desc)->getCLEntityType() == CL_ENTITY_TYPE::MESH)
```

Comment:
```suggestion
		if (shiftedObject->isKindOf(CC_TYPES::MESH))
```

Reply:
shiftedObject right here is the vertices cloud, will it still be mesh object?